provider/terraform/resources/resource_storage_default_object_acl.go

```diff
@@ -2,185 +2,109 @@ package google
+				},
 			},
 		},
 	}
 }
 
-func resourceStorageDefaultObjectAclCreate(d *schema.ResourceData, meta interface{}) error {
+func resourceStorageDefaultObjectAclCreateUpdate(d *schema.ResourceData, meta interface{}) error {
 	config := meta.(*Config)
 
 	bucket := d.Get("bucket").(string)
-	roleEntity := d.Get("role_entity").([]interface{})
-
-	for _, v := range roleEntity {
-		pair, err := getRoleEntityPair(v.(string))
-
-		ObjectAccessControl := &storage.ObjectAccessControl{
+	defaultObjectAcl := []*storage.ObjectAccessControl{}
+	for _, v := range d.Get("role_entity").(*schema.Set).List() {
+		pair := getValidatedRoleEntityPair(v.(string))
+		defaultObjectAcl = append(defaultObjectAcl, &storage.ObjectAccessControl{
 			Role:   pair.Role,
 			Entity: pair.Entity,
-		}
-
-		log.Printf("[DEBUG]: setting role = %s, entity = %s on bucket %s", pair.Role, pair.Entity, bucket)
-
-		_, err = config.clientStorage.DefaultObjectAccessControls.Insert(bucket, ObjectAccessControl).Do()
-
-		if err != nil {
-			return fmt.Errorf("Error setting Default Object ACL for %s on bucket %s: %v", pair.Entity, bucket, err)
-		}
+		})
 	}
-	d.SetId(bucket)
-	return resourceStorageDefaultObjectAclRead(d, meta)
-}
-
-func resourceStorageDefaultObjectAclRead(d *schema.ResourceData, meta interface{}) error {
-	config := meta.(*Config)
-
-	bucket := d.Get("bucket").(string)
-
-	roleEntities := make([]interface{}, 0)
-	reLocal := d.Get("role_entity").([]interface{})
-	reLocalMap := make(map[string]string)
-	for _, v := range reLocal {
-		res, err := getRoleEntityPair(v.(string))
-
-		if err != nil {
-			return fmt.Errorf(
-				"Old state has malformed Role/Entity pair: %v", err)
-		}
-
-		reLocalMap[res.Entity] = res.Role
-	}
-
-	res, err := config.clientStorage.DefaultObjectAccessControls.List(bucket).Do()
 
+	res, err := config.clientStorage.Buckets.Get(bucket).Do()
 	if err != nil {
-		return handleNotFoundError(err, d, fmt.Sprintf("Storage Default Object ACL for bucket %q", d.Get("bucket").(string)))
+		return fmt.Errorf("Error reading bucket %s: %v", bucket, err)
 	}
 
-	for _, v := range res.Items {
-		role := v.Role
-		entity := v.Entity
-		// We only store updates to the locally defined access controls
-		if _, in := reLocalMap[entity]; in {
-			roleEntities = append(roleEntities, fmt.Sprintf("%s:%s", role, entity))
-			log.Printf("[DEBUG]: saving re %s-%s", v.Role, v.Entity)
+	// Even with ForceSendFields the empty array wasn't working. Luckily, this is the same thing
+	if len(defaultObjectAcl) == 0 {
+		_, err = config.clientStorage.Buckets.Update(bucket, res).PredefinedDefaultObjectAcl("private").Do()
+		if err != nil {
+			return fmt.Errorf("Error updating default object acl to empty for bucket %s: %v", bucket, err)
+		}
+	} else {
+		res.DefaultObjectAcl = defaultObjectAcl
+		_, err = config.clientStorage.Buckets.Update(bucket, res).Do()
+		if err != nil {
+			return fmt.Errorf("Error updating default object acl for bucket %s: %v", bucket, err)
```

Comment:
Is it normal to set the id in the create method after success? I know that read will also set it, but I've seen this in many other resources.

Reply:
It is normal to! This is because if there's an error between Create and the Read at the bottom of Create, Terraform won't know that the resource was created and the user will need to delete it upstream or import it. Since this is a fine-grained resource that always exists (when the parent does), we don't need Terraform to do that - we can just run another Create.

If you feel that we should add a `SetId` anyways, lmk and I can - my preference here is to omit it.

Reply:
No strong preference here, just curious.